File: bestia-client/src/Bnet/Message/Entity/DeadComponentSMSG.cs
using Godot;

namespace BestiaBehemothClient.Bnet.Message.Entity
{
  /// <summary>
  /// A player-owned entity is lying dead where it fell, awaiting a respawn. Unlike a wild mob - which
  /// is destroyed outright and arrives as a VanishEntitySMSG of kind DEATH - a player body stays in
  /// the world, so this is entity state and everyone in range receives it. Getting back up arrives as
  /// this same message with Removed = true.
  /// </summary>
  [GlobalClass]
  public partial class DeadComponentSMSG : EntitySMSG
  {
    [Export] public bool Removed { get; set; }

    public static DeadComponentSMSG FromProto(global::Bnet.DeadComponentSMSG proto)
    {
      return new DeadComponentSMSG()
      {
        EntityId = proto.EntityId,
        Removed = proto.Removed
      };
    }

    public override string ToString()
    {
      return $"DeadComponentSMSG(EntityId={EntityId}, Removed={Removed})";
    }
  }
}

File: zone-server/src/main/kotlin/net/bestia/zone/ecs/battle/damage/DeadComponentSMSG.kt
package net.bestia.zone.ecs.battle.damage

import net.bestia.bnet.proto.DeadComponentSmsgProto
import net.bestia.bnet.proto.EnvelopeProto
import net.bestia.zone.message.EntitySMSG

/**
 * Tells everyone in range that a player-owned entity is lying dead, and - with [removed] - that it is
 * back on its feet. Produced by [Dead.toEntityMessage].
 */
data class DeadComponentSMSG(
  override val entityId: Long,
  val removed: Boolean = false
) : EntitySMSG {

  override fun toBnetEnvelope(): EnvelopeProto.Envelope {
    val proto = DeadComponentSmsgProto.DeadComponentSMSG.newBuilder()
      .setEntityId(entityId)
      .setRemoved(removed)
      .build()

    return EnvelopeProto.Envelope.newBuilder()
      .setCompDead(proto)
      .build()
  }
}
